Regular grid interpolation

flabowski opened the pull request:
see #243

The class needs a grid, i.e. something like `grid = ([0.1, 0.2, 0.3], [1, 2, 3, 4, 5])`, not a list of points such as `points = [[0.1, 1], [0.1, 2], [0.1, 3], ..., [0.3, 5]]`. That means [rom.fit()](https://github.com/mathLab/EZyRB/blob/8a51f55e08d520e2c661d57c67a99e03dc1253ee/ezyrb/reducedordermodel.py#L67) does not work, the method assumes we want to use the points from the database to construct the interpolator. Instead the user has to call `reduction.fit(...)` first and then `rg_approximation.fit(...)` with the proper grid.
Although the interpolation returns what we expect in the tests, I have not tested it in a full ROM yet.

mtezzele commented:
Hi @flabowski. What about adding a preprocessing step within this class to create the 1d vectors from the parameters given by the `Database` class? The check on the inputs should raise an error if there are not on a grid, and depending on the dimensions create the vectors to feed the `RegularGridInterpolator`. What do you think?

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,76 @@
+        """
+        Construct the interpolator given `points` and `values`.
+
+        Parameters
```
Please use the same documentation style you find in the rest of the package.

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+    """
+    Multidimensional interpolator on regular grids.
+    See scipy's RegularGridInterpolator
+
```
Please add the class documentation with the members description and possibly a minimal example. See [here](https://github.com/mathLab/EZyRB/blob/8a51f55e08d520e2c661d57c67a99e03dc1253ee/ezyrb/gpr.py#L15) for reference.

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        a tuple of ndarray of float, with shapes (m1, ), …, (mn, )
+        The values are ordered so they fit on a mesh generated with
+        numpy.meshgrid(ax1, ax2, ..., axn, indexing="ij")
+
```
Add a properly formatted documentation.

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+            shape = []
+        for i in range(self.dim):
+            shape.append(len(grid_axes[i]))
+        assert np.prod(shape) == values_grd.size, "Values don't match grid. "\
```
Add a ValueError

flabowski replied:
I moved it entirely, I thought here is not the right place since get_grid_axes() has to ensure this

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        :param array_like points: the coordinates of the points.
+        :param array_like values: the values in the points.
+        """
+        # we have two options
```
Remove unnecessary comments

fandreuz replied:
Some high level implementation details could be moved to a small paragraph in the class documentation

flabowski replied:
The improved docstrings and the code should be explanatory now

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+                                                        shape),
+                                                    **kvargs)
+
+    def predict(self, new_point):
```
Add documentation

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        if len(new_point.shape) == 1:
+            new_point.shape = (-1, 1)
+
+        dim = self.dim
```
no need to use a new variable

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+    :param list iN: indices representing the position of the element(s)
+                    in each dimension.
+    :param list nN: size of the array in each dimension.
+
```
Add `:return: description of the returned values.`

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+            xi_extended = new_point
+        return self.interpolator(xi_extended).T
+
+
```
I suggest moving this method inside the class. Make it private with an underscore as first character, and add the @staticmethod decorator on top of it.

flabowski replied:
I removed it entirely since it is implemented in numpy already... Thanks @fandreuz for the hint

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        return self.interpolator(xi_extended).T
+
+
+def calculate_flat_index(iN, nN):
```
use meaningful variable names. No camelcase please.

mtezzele commented on tests/test_regular_grid.py:
```diff
@@ -0,0 +1,96 @@
+        assert rom.predict([1]) == 1
+        assert rom.predict([2]) == 5
+        assert rom.predict([3]) == 3
+    # TODO: test kvargs? depend on scipy version....
```
you can remove this comment. I think the tests are enough.

mtezzele reviewed (commented):
Could you please add a doc file for the automatic doc generation of the new module. You can follow what has been done in `docs/source/rbf.rst` for example. Then just add the name of the new file here: `docs/source/code.rst`

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+            iN.append(unique_inverse_n)
+
+        if np.prod(nN) != len(vals_scrmbld):
+            raise ValueError("points and values are not on a regular grid")
```
I'm not sure this error is the best we can do. As a user I'd like to get some more info on the cause of the error. I propose something like:
```suggestion
            raise ValueError("Points and values are not laid out on a regular grid, the size of vals_scrmbld is not equal to ...")
```
and replacing the dots with something meaningful.

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+            raise ValueError("points and values are not on a regular grid")
+        new_row_index = calculate_flat_index(iN, nN)
+        reverse_scrambling = np.argsort(new_row_index)
+        vals_on_regular_grid = vals_scrmbld[reverse_scrambling, :]
```
```suggestion
        vals_on_regular_grid = vals_scrmbld[reverse_scrambling]
```

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        # or 2.: we "interpolate" the mode number
+        # option 1 is cleaner, but option 2 performs better
+        # X = U S VT, X being shaped (m, n)
+        points = np.array(points)
```
```suggestion
        points = np.asarray(points)
```

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+            raise ValueError('Invalid format or dimension for the argument'
+                             '`points`.')
+
+        if len(points.shape) == 1:
+            points.shape = (-1, 1)
```
```suggestion
        points = np.atleast_2d(points)
```

flabowski replied:
That lets the test test_predict1d_2() fail.
I assumed x1 = [1, 2] are 2 1D points. The old version results thus in array([[1], [2]]), what you suggested results in array([[1, 2]]), which is 1 point in 2 dimensions. Since the Linear class interprets an input like x1 as 2 points in 1D, I suggest to revert that change to the behaviour I originally implemented.

fandreuz replied:
Thanks for clarifying the situation, anyway could you please consider this change?
```suggestion
        if points.ndim == 1:
            points = points[:, None]
```
It feels more idiomatic to me and it follows the same style we use in other parts of PyDMD. If you agree with it, please apply the same to `new_point` below.

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        if len(points.shape) == 1:
+            points.shape = (-1, 1)
+
+        self.dim = len(points[0])
```
```suggestion
        self.dim = points.shape[1]
```

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        self.dim = len(points[0])
+        vals = np.asarray(values)
+        grid_axes, values_grd = self.get_grid_axes(points, vals)
+        self.n_modes = vals.T.shape[0]  # vals = (S@VT).T = S@V
```
```suggestion
        self.n_modes = vals.shape[-1]  # vals = (S@VT).T = S@V
```

fandreuz replied:
Not clear what is `S` in the comment, could you please clarify?

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        vals_on_regular_grid = vals_scrmbld[reverse_scrambling, :]
+        return grid_axes, vals_on_regular_grid
+
+    def fit(self, points, values, **kvargs):
```
Is there a reason why you don't use the standard `kwargs` here?
```suggestion
    def fit(self, points, values, **kwargs):
```

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+                                                    **kvargs)
+
+    def predict(self, new_point):
+        new_point = np.array(new_point)
```
```suggestion
        new_point = np.asarray(new_point)
```

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+
+    def predict(self, new_point):
+        new_point = np.array(new_point)
+        if len(new_point.shape) == 1:
+            new_point.shape = (-1, 1)
```
```suggestion
        new_point = np.atleast_2d(new_point)
```

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        dim = self.dim
+        if self.n_modes > 1:
+            xi_extended = np.zeros((len(self.mode_nr), len(new_point), dim+1))
+            xi_extended[:, :, 0] = self.mode_nr[:, None]
+            for i in range(dim):
+                xi_extended[:, :, i+1] = np.array(new_point)[:, i]
```
```suggestion
            xi_extended[..., 0] = self.mode_nr[:, None]
            xi_extended[..., 1:] = new_point
```

fandreuz commented on ezyrb/regular_grid.py:
```diff
@@ -0,0 +1,131 @@
+        return self.interpolator(xi_extended).T
+
+
+def calculate_flat_index(iN, nN):
```
No need to implement this function, is already provided by NumPy: https://numpy.org/doc/stable/reference/generated/numpy.ravel_multi_index.html#numpy.ravel_multi_index

It provides also error handling for out of bound indices

flabowski replied:
oops, did know that. thanks!

fandreuz reviewed (commented):
Hi @flabowski, I left some comments on your code. Let me know if you need any clarification. Thanks!

flabowski commented:
Hei @mtezzele  and @fandreuz,
Thank you for your feedback, great to see that there is some interest in this PR :) I think i adressed all of your comments, let me know in case I missed something.

flabowski commented:
I realized now that one of the changes @fandreuz suggested changed the behaviour and one of the tests failed. I assume that was unintended. I reverted it, so the behaviour matches the `Linear` interpolator again. Also, I introduced a simplification, which should not change the output.

mtezzele commented on ezyrb/regular_grid.py:
```diff
@@ -55,77 +91,27 @@ def fit(self, points, values, **kvargs):
+        """
+        new_point = np.asarray(new_point)
         if len(new_point.shape) == 1:
             new_point.shape = (-1, 1)
```
please fix this as above

mtezzele reviewed (commented):
Please add the new rst doc file in the following file: `docs/source/code.rst`

ndem0 commented on tests/test_regular_grid.py:
```diff
@@ -0,0 +1,95 @@
+        rom.fit()
+        assert rom.predict([1]) == 1
+        assert rom.predict([2]) == 5
+        assert rom.predict([3]) == 3
```
Please add another test in order to check that the exception is correctly raised in case the passed points are not a grid (see https://github.com/mathLab/EZyRB/blob/8a51f55e08d520e2c661d57c67a99e03dc1253ee/tests/test_radius_neighbors_regressor.py#L68 for an example)

ndem0 commented:
Dear @flabowski, thanks for the effort! From my side, everything looks almost perfect, I just asked if you can add a simple test in order to check that the exception is raised in case of not grid points.

mtezzele commented on docs/source/regular_grid.rst:
```diff
@@ -10,10 +10,10 @@ RegularGrid
     :nosignatures:
 
     RegularGrid
+    RegularGrid.def
```
This is not needed. Or at least we don't use it in the other files. I am not sure about the generated outcome...

flabowski replied:
I am not sure either. The file was generated automatically when ran `make_rst.sh`. Some other things seem to be fishy, I am not sure how to fix that. E.g. `EZyRB/docs/build/html/code.html` lists the new module on the left under **Code Documentation** and on the right, but the class members are missing on the right. Also, when I click on any other module (e.g. **Code documentation -> linear**), the new module disappears on the left under **Code Documentation**. When I click on **Code documentation -> regular_grid** i.e. `EZyRB/docs/build/html/regular_grid.html`, the whole left side menu appears to be collapsed.
I dont really know if that is only on my end or if these are problems for other users too? Do you have a suggestion how to fix that? Maybe @ndem0 knows?
Another thing is that there seem to be two .rst files for the same module, `podae.rst` and `pod_ae.rst`. One appears to be autogenerated, the other one hand written. Quite a few of the .rst files seem to be different (outdated?) from what is generated when i run `make_rst.sh`. Maybe it is something for a new issue?

ndem0 replied:
We are not using that script anymore! Sorry, it's not written, I gonna remove it in the next release of EZyRB (#232).
The file now looks correct, thanks!
